flexmeasures/data/models/planning/storage.py

```diff
@@ -85,46 +89,64 @@ def compute(self, skip_validation: bool = False) -> pd.Series | None:
         self.sensor.check_required_attributes([("capacity_in_mw", (float, int))])
 
         # Check for known prices or price forecasts, trimming planning window accordingly
-        up_deviation_prices, (start, end) = get_prices(
-            (start, end),
-            resolution,
-            beliefs_before=belief_time,
-            price_sensor=consumption_price_sensor,
-            sensor=sensor,
-            allow_trimmed_query_window=False,
-        )
-        down_deviation_prices, (start, end) = get_prices(
-            (start, end),
-            resolution,
-            beliefs_before=belief_time,
-            price_sensor=production_price_sensor,
-            sensor=sensor,
-            allow_trimmed_query_window=False,
-        )
+        up_deviation_prices_array = []
+        for power_sensor, price_sensor in consumption_price_sensor_per_device.items():
```

Comment:
For example, if the `StorageScheduler` is called through the api, it will fail.

Reply:
Agreed. Shall we just use if case to handle that ?

Reply:
Alternatively, I was wondering if I could append the 'consumption_price_sensor' to 'consumption_price_sensor_per_device' ,and similarly for production, by making the dictionary link between 'sensor=self.sensor' with 'consumption_price_sensor'?
What do you think about that?If you agree I will proceed with that.

Reply:
Yes, I think that could work. If `consumption_price_sensor_per_device` is not provided, you can create internally one that maps each device to the same `consumption_price_sensor`. The same for production.

Reply:
Do I need to map each device with `consumption_price_sensor` or could i just map `consumption_price_sensor` with the battery sensor either?Shouldn't that work as well?

Reply:
>  If consumption_price_sensor_per_device is not provided, you can create internally one that maps each device to the same consumption_price_sensor. The same for production.

This would not be the same as applying the consumption price sensor to the aggregate power flow, which is the use case we currently support.

For example, for a given time step, we currently have:

*Example A*

- Device A consumes 10 kWh
- Device B produces 8 kWh
- Devices A and B belong to the same system, with an aggregate consumption of 2 kWh
- The system's aggregate consumption is priced at 0.5 EUR/kWh
- The system costs are 1 EUR.

If I am reading your suggestion right, you propose:

*Example B*

- Device A consumes 10 kWh, priced at 0.5 EUR/kWh, incurring costs of 5 EUR
- Device B produces 8 kWh, priced at 0.5 EUR/kWh, incurring a revenue of 4 EUR
- The system costs are 5 - 4 = 1 EUR

Is that what you had in mind?

This might look the same, but things change when the consumption price differs from the production price. Let's say the production price is 0.4 EUR/kWh. Example A wouldn't change, because there is no aggregate production, only aggregate consumption. Example B changes, however:

- Device A consumes 10 kWh, priced at 0.5 EUR/kWh, incurring costs of 5 EUR
- Device B produces 8 kWh, priced at 0.4 EUR/kWh, incurring a revenue of 3.20 EUR
- The system costs are 5 - 3.20 = 1.80 EUR

In my opinion, we should not touch the current use case by trying to map the `consumption_price_sensor` argument to individual devices. Instead, we should extend the cost function with *additional* cost components for each device, allowing to specify a consumption and production price per device. If a price for a specific device is not specified, it can be assumed to be zero.

Reply:
Yes agreed @victorgarcia98 .But what I was suggesting was to map `consumption_price_sensor` to the battery sensor(which would in turn take care of the aggregate consumption/production just like in Example A). In my opinion ,that's gonna work perfectly as mapping it with the battery would mean mapping it up with the system.Let me know if that's something that works?

Reply:
I don't follow. Can you explain it with an example in more detail, with example values?

Reply:
Earlier, we had the following:
```up_deviation_prices, (start, end) = get_prices(
            (start, end),
            resolution,
            beliefs_before=belief_time,
            price_sensor=consumption_price_sensor,
            sensor=sensor,
            allow_trimmed_query_window=False,
        )```
which now has changed to this:
```up_deviation_prices_array = []
        for power_sensor, price_sensor in consumption_price_sensor_per_device.items():
            up_deviation_prices, (start, end) = get_prices(
                (start, end),
                resolution,
                beliefs_before=belief_time,
                price_sensor=price_sensor,
                sensor=power_sensor,
                allow_trimmed_query_window=False,
            )
            up_deviation_prices_array.append(up_deviation_prices)```
Keeping this in mind, I believe the sensor to be passed to `get_prices`  should be the battery power sensor to make sure the price sensor is applied to the aggregate power flow.
So the following code should be added to `compute` funtion :
```#Convert single price sensors to Multiple Price Sensors Dict
        if consumption_price_sensor is not None:
            consumption_price_sensor_per_device[sensor]=consumption_price_sensor
        
        if production_price_sensor is not None:
            production_price_sensor_per_device[sensor]=production_price_sensor```
This will make the situation exactly like Example A.There is no difference in the example.